src/app/store/evm/assets/reducer.ts

```diff
@@ -0,0 +1,78 @@
+import { createReducer } from '@reduxjs/toolkit';
+import { getAddress } from 'viem';
+
+import { NATIVE_TOKEN_INDEX } from 'lib/apis/temple/endpoints/evm/api.utils';
+import { toTokenSlug } from 'lib/assets';
+import { isPositiveCollectibleBalance, isPositiveTokenBalance } from 'lib/utils/evm.utils';
+
+import { proceedLoadedEvmAssetsAction, putNewEvmCollectibleAction, putNewEvmTokenAction } from './actions';
+import { EvmAssetsInitialState, EvmAssetsStateInterface } from './state';
+
+export const evmAssetsReducer = createReducer<EvmAssetsStateInterface>(EvmAssetsInitialState, builder => {
+  builder.addCase(proceedLoadedEvmAssetsAction, ({ tokens, collectibles }, { payload }) => {
+    const { publicKeyHash, chainId, data } = payload;
+
+    if (!tokens[publicKeyHash]) tokens[publicKeyHash] = {};
+    const accountTokens = tokens[publicKeyHash];
```

Comment:
`get` does not crash if a top-level object does not exist but returns `undefined`, and `set` creates a nested object if necessary if a key is not a number. However, `chainId` is a number, so `set` will create an array, which is bullshit in our case. Well, let's then create a function like
```
function ensureObject<K extends string | number, V>(parent: Record<K, StringRecord<V>>, key: K): StringRecord<V>;
function ensureObject<K extends string | number, V>(parent: Record<K, Record<number, V>>, key: K): Record<number, V>;
function ensureObject<K extends string | number, V>(parent: Record<K, Record<string | number, V>>, key: K) {
  if (!parent[key]) {
    parent[key] = {};
  }

  return parent[key];
}
```
and use it here to avoid logic duplication.

Reply:
wrote more specific solution, pls take a look